File: algo/lc.387.cc
class Solution {
public:
    int firstUniqChar(string s) {
        map<char, int> counter;
        // create dictionary
        for (auto i : s) {
            auto cursor = counter.find(i);
            if (cursor != counter.end()) {
                cursor->second += 1;
            } else {
                counter.insert(pair<char, int>(i, 1));
            }
        }
        // scan
        for (int i = 0; i < s.size(); i++) {
            auto cursor = counter.find(s[i]);
            if (cursor->second == 1)
                return i;
        }
        return -1;
    }
};

File: algo/lc.387.py
class Solution:
    def firstUniqChar(self, s: str) -> int:
        d = {}
        for char in s:
            if char in d:
                d[char] += 1
            else:
                d[char] = 1
        for (idx, char) in enumerate(s):
            if d[char] == 1:
                return idx
        return -1
